kolibri/core/logger/upgrade.py

```diff
@@ -13,3 +16,31 @@ def migrate_examlog_and_examattemptlogs():
     data structures.
     """
     migrate_from_exam_logs(ExamLog.objects.all())
+
+
+@version_upgrade(old_version="<0.15.11")
+def fix_asymptotic_progress():
+    """
+    Migrate any ContentSummaryLogs that have erroneously had their progress set to >0.999 but not to 1
+    """
+    for summarylog in ContentSummaryLog.objects.filter(
+        progress__gt=0.999, progress__lt=1
+    ):
+        # Do this in a for loop rather than trying to minimize write queries, as it's a one time event
```

Comment:
Do upgrade routines get run within a transaction? May help speed it up.

Reply:
No - I also assume that the number of writes here will generally be quite small, so was not optimizing for speed.